src/large_pages/node_large_page.cc

```diff
@@ -110,72 +111,56 @@ inline uintptr_t hugepage_align_down(uintptr_t addr) {
   return ((addr) & ~((hps) - 1));
 }
 
-// The format of the maps file is the following
-// address           perms offset  dev   inode       pathname
-// 00400000-00452000 r-xp 00000000 08:02 173521      /usr/bin/dbus-daemon
-// This is also handling the case where the first line is not the binary.
+struct dl_iterate_params {
+  uintptr_t start;
+  uintptr_t end;
+  uintptr_t reference_sym;
+};
 
 struct text_region FindNodeTextRegion() {
   struct text_region nregion;
   nregion.found_text_region = false;
 #if defined(__linux__)
-  std::ifstream ifs;
-  std::string map_line;
-  std::string permission;
-  std::string dev;
-  char dash;
-  uintptr_t start, end, offset, inode;
-  uintptr_t node_text_start = reinterpret_cast<uintptr_t>(&__node_text_start);
+  dl_iterate_params dl_params = {
+    0, 0, reinterpret_cast<uintptr_t>(&__node_text_start)
+  };
   uintptr_t lpstub_start = reinterpret_cast<uintptr_t>(&__start_lpstub);
 
-  ifs.open("/proc/self/maps");
-  if (!ifs) {
-    PrintWarning("could not open /proc/self/maps");
-    return nregion;
-  }
-
-  while (std::getline(ifs, map_line)) {
-    std::istringstream iss(map_line);
-    iss >> std::hex >> start;
-    iss >> dash;
-    iss >> std::hex >> end;
-    iss >> permission;
-    iss >> offset;
-    iss >> dev;
-    iss >> inode;
-
-    if (inode == 0)
-      continue;
-
-    std::string pathname;
-    iss >> pathname;
-
-    if (permission != "r-xp")
-      continue;
-
-    if (node_text_start < start || node_text_start >= end)
-      continue;
-
-    start = node_text_start;
-    if (lpstub_start > start && lpstub_start <= end)
-      end = lpstub_start;
-
-    char* from = reinterpret_cast<char*>(hugepage_align_up(start));
-    char* to = reinterpret_cast<char*>(hugepage_align_down(end));
-
-    if (from >= to)
-      break;
+  if (dl_iterate_phdr([](struct dl_phdr_info* info, size_t, void* data) -> int {
+    if (info->dlpi_name[0] == 0) {
+      for (int idx = 0; idx < info->dlpi_phnum; idx++) {
+        const ElfW(Phdr)* phdr = &info->dlpi_phdr[idx];
+        if (phdr->p_type == PT_LOAD && phdr->p_flags & PF_X) {
+          auto dl_params = reinterpret_cast<dl_iterate_params*>(data);
+          uintptr_t start = info->dlpi_addr + phdr->p_vaddr;
+          uintptr_t end = start + phdr->p_memsz;
+
+          if (dl_params->reference_sym >= start &&
+              dl_params->reference_sym <= end) {
```

Comment:
Just checking, is `end` inclusive or exclusive?

Aside: it looks like there's no real need for the `dl_params->reference_sym` field. This also works, wouldn't it?
```c++
auto reference_sym = reinterpret_cast<uintptr_t>(&__node_text_start);
if (reference_sym >= start && reference_sym <= end) {
  // ...
}
```

Reply:
@bnoordhuis the `reference_sym` field is being used both inside the `dl_iterate_phdr` callback and outside of it. I thought it best not to assign `reinterpret_cast<uintptr_t>(&__node_text_start);` to two different variables because, if the symbol changes in the future, they might get out of sync.